### PR TITLE
Initial Windows scancode implementation

### PR DESCRIPTION
**Pull request by JonnyPtn**
**SFML Tasks**

* [ ] Adjust code, according to the review comments
* [ ] Merge into the feature/scancode branch

---

Just the events side of things so far. I'm away for a few days now but feel free to pick it apart before I return.

Also made me dig up #304 as raw input might be worth considering for keyboard input too?

### CODE REVIEW

**Comment by eXpl0it3r** [comment]
> Also made me dig up #304 as raw input might be worth considering for keyboard input too?

Let's not mix topics unless it absolutely makes sense. Every OS specific feature, needs again to be implemented on the other OS. :wink:

**Comment by JonnyPtn** [comment]
If you try helping with this you'll see why I thought it made sense to mention it - Scancodes are a bitch on windows, especially when using a legacy API

also it's windows specific so although other platforms may have options to implement raw input, I'm talking specifically about windows, where rawinput would help with implementing scancodes

**Comment by JonnyPtn** [comment]
This looks to be working well with test program mentioned in #1235, could do with some testing (Once I get hold of a keyboard with the extra buttons I'll try and fill in the blanks)

File: src/SFML/Window/Win32/InputImpl.cpp [hunk 1]
```diff
@@ -48,113 +48,149 @@ bool InputImpl::isKeyPressed(Keyboard::Key key)
...
-        case Keyboard::F7:         vkey = VK_F7;         break;
-        case Keyboard::F8:         vkey = VK_F8;         break;
-        case Keyboard::F9:         vkey = VK_F9;         break;
-        case Keyboard::F10:        vkey = VK_F10;        break;
-        case Keyboard::F11:        vkey = VK_F11;        break;
-        case Keyboard::F12:        vkey = VK_F12;        break;
-        case Keyboard::F13:        vkey = VK_F13;        break;
-        case Keyboard::F14:        vkey = VK_F14;        break;
-        case Keyboard::F15:        vkey = VK_F15;        break;
-        case Keyboard::Pause:      vkey = VK_PAUSE;      break;
```
Comment (eXpl0it3r) [inline src/SFML/Window/Win32/InputImpl.cpp]: Looks like your IDE auto "corrected" the indention here, unfortunately it's not how we format switch-cases.

File: src/SFML/Window/Win32/InputImpl.cpp [hunk 2]
```diff
@@ -48,113 +48,149 @@ bool InputImpl::isKeyPressed(Keyboard::Key key)
...
 
     return (GetAsyncKeyState(vkey) & 0x8000) != 0;
 }
 
+////////////////////////////////////////////////////////////
+bool InputImpl::isKeyPressed(Keyboard::Scancode code)
+{
+    auto winCode = sfScanToWin(code);
+    auto vkey = MapVirtualKey(winCode, MAPVK_VSC_TO_VK_EX);
+    auto state = GetAsyncKeyState(vkey);
```
Comment (eXpl0it3r) [inline src/SFML/Window/Win32/InputImpl.cpp]: `auto` is a C++11 feature, SFML 2 is still on C++03.

File: src/SFML/Window/Win32/InputImpl.cpp [hunk 3]
```diff
@@ -251,6 +287,148 @@ Vector2i InputImpl::getTouchPosition(unsigned int /*finger*/, const Window& /*re
...
+    case Keyboard::ScanLControl:            return 29;
+    case Keyboard::ScanLShift:              return 42;
+    case Keyboard::ScanLAlt:                return 56;
+    case Keyboard::ScanLSystem:             return 91 | KF_EXTENDED ;
+    case Keyboard::ScanRControl:            return KF_EXTENDED | 29;
+    case Keyboard::ScanRShift:              return 54;
+    case Keyboard::ScanRAlt:                return 56;
+    //case Keyboard::ScanRSystem:           return ? ? ? ;
+
+    default: return 0; // Not sure what to return here?
```
Comment (eXpl0it3r) [inline src/SFML/Window/Win32/InputImpl.cpp]: Wrong switch-case indentation.

File: src/SFML/Window/Win32/InputImpl.hpp [hunk 4]
```diff
@@ -31,6 +31,7 @@
 #include <SFML/Window/Keyboard.hpp>
 #include <SFML/Window/Mouse.hpp>
 
+#include <windows.h>
```
Comment (eXpl0it3r) [inline src/SFML/Window/Win32/InputImpl.hpp]: Is this needed to get access to `WORD`?

File: src/SFML/Window/Win32/InputImpl.hpp [hunk 5]
```diff
@@ -176,6 +177,10 @@ class InputImpl
     ///
     ////////////////////////////////////////////////////////////
     static Vector2i getTouchPosition(unsigned int finger, const Window& relativeTo);
+
+    private:
+
+    static WORD sfScanToWin(Keyboard::Scancode code);
```
Comment (eXpl0it3r) [inline src/SFML/Window/Win32/InputImpl.hpp]: The sf in `sfScanToWin` seems kind of odd, i.e. making the namespace a prefix. I'd prefer a different name.

File: src/SFML/Window/Win32/WindowImplWin32.cpp [hunk 6]
```diff
@@ -559,6 +559,107 @@ void WindowImplWin32::grabCursor(bool grabbed)
...
+    case 80: return (HIWORD(flags) & KF_EXTENDED) ? Keyboard::ScanDown : Keyboard::ScanNumpad2;
+    case 81: return (HIWORD(flags) & KF_EXTENDED) ? Keyboard::ScanPageDown : Keyboard::ScanNumpad3;
+    case 82: return (HIWORD(flags) & KF_EXTENDED) ? Keyboard::ScanInsert : Keyboard::ScanNumpad0;
+    case 83: return (HIWORD(flags) & KF_EXTENDED) ? Keyboard::ScanDelete : Keyboard::ScanDecimal;
+
+    case 91: return (HIWORD(flags) & KF_EXTENDED) ? Keyboard::ScanLSystem : Keyboard::ScanUnknown;
+    case 93: return (HIWORD(flags) & KF_EXTENDED) ? Keyboard::ScanMenu : Keyboard::ScanUnknown;
+
+    default: return Keyboard::ScanUnknown;
+    }
```
Comment (eXpl0it3r) [inline src/SFML/Window/Win32/WindowImplWin32.cpp]: Wrong switch-case indentation.

File: src/SFML/Window/Win32/InputImpl.cpp (hunk 2 above)
Comment (eliasdaler) [inline src/SFML/Window/Win32/InputImpl.cpp]: I'm a bit confused by what's happening here. Can't we map Keyboard::Scancode directly to virtual keys? (we can get them from events, no need to go to Windows scancodes and back)

File: src/SFML/Window/Win32/WindowImplWin32.cpp [hunk 7]
```diff
@@ -732,7 +833,8 @@ void WindowImplWin32::processEvent(UINT message, WPARAM wParam, LPARAM lParam)
                 event.key.control = HIWORD(GetAsyncKeyState(VK_CONTROL)) != 0;
                 event.key.shift   = HIWORD(GetAsyncKeyState(VK_SHIFT))   != 0;
                 event.key.system  = HIWORD(GetAsyncKeyState(VK_LWIN)) || HIWORD(GetAsyncKeyState(VK_RWIN));
-                event.key.code    = virtualKeyCodeToSF(wParam, lParam);
+                event.key.scancode = toScancode(lParam);
+                event.key.code = virtualKeyCodeToSF(wParam, lParam);
```
Comment (eliasdaler) [inline src/SFML/Window/Win32/WindowImplWin32.cpp]: I'm a bit confused by what's happening here. Can't we map scancode that came from windows event to virtual key and then only use it for scancode implementation? I think it's better to map virtual key codes to sf::Keyboard::Scancode than take specific scancodes into account (those are hardware specific, AFAIK)

**Comment by eliasdaler** [comment]
Posted on a forum, but I'll post it here too:

Bugs found:

On QWERTY:

* Tilde/Grave accent is identified as ScanBackslash
* F11 and F12 don't have a string description
* Backslash (the key left to Enter) is detected as ScanDash and getDescription returns "`"
* Volume Up/Down, Find and Mute buttons are "Unknown"
* Keyboard enter is "ScanReturn", though it should be "ScanEnter". Numpad enter should be "ScanReturn" (see comments in enum)

Other keyboard layouts work, but I wonder if "ScanReverseSolidus" will be identified correctly. Someone who has a keyboad with this key (it's the one right to the LShift, it usually has "<>|" on it should test it.

* And getDescription returns translated descriptions for Ctrl, Shift, Alt, Home, PageUp, etc. in non-English layouts. E.g. if I turn on AZERTY, I get descriptions in French. Not sure if it's a behaviour we want... We'll discuss it in Scancode PR (#1235).

File: src/SFML/Window/Win32/InputImpl.cpp (hunk 1 above)
Comment (JonnyPtn) [reply src/SFML/Window/Win32/InputImpl.cpp]: Should be fixed now

File: src/SFML/Window/Win32/InputImpl.hpp (hunk 4 above)
Comment (JonnyPtn) [reply src/SFML/Window/Win32/InputImpl.hpp]: There's no need for it to be declared in the header, so I've removed it and the include now

File: src/SFML/Window/Win32/InputImpl.cpp (hunk 2 above)
Comment (JonnyPtn) [reply src/SFML/Window/Win32/InputImpl.cpp]: Fixed 👍

File: src/SFML/Window/Win32/InputImpl.cpp (hunk 3 above)
Comment (JonnyPtn) [reply src/SFML/Window/Win32/InputImpl.cpp]: fixed 👍

File: src/SFML/Window/Win32/WindowImplWin32.cpp (hunk 6 above)
Comment (JonnyPtn) [reply src/SFML/Window/Win32/WindowImplWin32.cpp]: fixed 👍

**Comment by JonnyPtn** [comment]
I've rebased, addressed some of the feedback and squashed

**Comment by eliasdaler** [comment]
Nice. Will review/test soon, great to see that this is alive. :)

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Pull request merged**